Create cloudflare/workers/backblaze.js

ibuclaw opened the pull request:
Generic cloudflare worker script that is able to treat a backblaze bucket as a static content website, used for both docarchives.dlang.io and downloads.dlang.org.

The only part that differs between the two is the `baseURL`, which I've redacted, as this can be information retrieved from GH actions when deploying the script.

ibuclaw committed:
Create cloudflare/workers/backblaze.js

Generic cloudflare worker script that is able to treat a backblaze bucket as a static content website, used for both docarchives.dlang.io and downloads.dlang.org.

CyberShadow commented on cloudflare/workers/backblaze.js:
```diff
@@ -0,0 +1,58 @@
+async function handleRequest(event) {
+  // only allow get requests
+  if (event.request.method !== 'GET') {
+    return new Response('Method not allowed', { status: 405 })
```
Nitpicking, but per spec 405 "MUST" specify which methods are allowed.

CyberShadow commented on cloudflare/workers/backblaze.js:
```diff
@@ -0,0 +1,58 @@
+  const response = await b2Fetch(path)
+  // all is well, return the response
+  if (response.status < 400) {
+    event.waitUntil(cache.put(event.request, response.clone()))
```
Nice, so it writes to the cache asynchronously but returns the response immediately. Hence `waitUntil` intead of `await`

ibuclaw replied:
Yes, https://developers.cloudflare.com/workers/runtime-apis/fetch-event#waituntil

CyberShadow commented on cloudflare/workers/backblaze.js:
```diff
@@ -0,0 +1,58 @@
+  }
+  if (response.status > 399) {
+    // return minimal error page
+    return new Response(response.statusText, { status: response.status })
```
Why construct a new response here? Is the returned error page liable to contain private information (like the bucket address)?

ibuclaw replied:
A default backblaze response is in JSON
```
{
  "code": "not_found",
  "message": "File with such name does not exist.",
  "status": 404
}
```
If there was a 404 page in the static pages, I would prefer to instead redirect to that instead.

CyberShadow reviewed (approved):
> Generic cloudflare worker script that is able to treat a backblaze bucket as a static content website, used for both docarchives.dlang.io and downloads.dlang.org.

Put this in `cloudflare/workers/README.md`?

ibuclaw committed:
backblaze.js: Add 'allowed' header in 405 response

CyberShadow commented on cloudflare/workers/backblaze.js:
```diff
@@ -9,7 +9,10 @@ addEventListener('fetch', event => {
-    return new Response('Method not allowed', { status: 405 })
+    return new Response('Method not allowed', {
+        status: 405,
+        headers: { 'allowed': 'GET' }
```
Isn't it `Allow`? Or is this some library-specific name?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Allow

ibuclaw replied:
Genuine typo on my part, the hands moved quicker than the brain.

ibuclaw committed:
Update backblaze.js

Fix typo in 405 response header.

ibuclaw committed:
Store backblaze base url as environment variable in cloudflare

ibuclaw committed:
Forward client headers to the B2 backend

So we can support range GET requests from the worker.  Don't override the response headers either, just use all that B2 gives us.
